Implement stereoisomer enumeration

greglandrum opened the pull request:
#### What does this implement/fix? Explain your changes.

Adds the ability to do stereochemistry expansion on molecules, including a heuristic to check whether or not the resulting structures are physically reasonable.

#### Any other comments?

Once we have some experience with the basics, moving this from Python to C++ is pretty straightforward.

@bp-kelley : since I know you've thought about this, your feedback on the approach would be really useful.

Relevant to #626

greglandrum commented:
Also: expanding this to support double bond stereochem turns out to be pretty easy. Is it worth doing?

greglandrum commented:
@coleb : I think we've talked about this before too, so you may also have some thoughts.

bp-kelley commented on rdkit/Chem/AllChem.py:
```diff
@@ -451,6 +450,102 @@ def AssignBondOrdersFromTemplate(refmol, mol):
+    nCenters = len(possibleCenters)
+    if not nCenters:
+        return (tm,)
+    bitflag = (1<<nCenters)-1
```
This could be seriously large :)

I suggest having a max nCenters, or a max output.

bp-kelley commented on rdkit/Chem/AllChem.py:
```diff
@@ -451,6 +450,102 @@ def AssignBondOrdersFromTemplate(refmol, mol):
+                tm.GetAtomWithIdx(possibleCenters[i]).SetChiralTag(CHI_TETRAHEDRAL_CW)
+        if tryEmbedding:
+            ntm = AddHs(tm)
+            cid = EmbedMolecule(ntm,randomSeed=bitflag)
```
randomSeed=bitflag is cute and deterministic :)

bp-kelley commented on rdkit/Chem/AllChem.py:
```diff
@@ -451,6 +450,102 @@ def AssignBondOrdersFromTemplate(refmol, mol):
+        else:
+            cid = 1
+        if cid>= 0:
+            res.append(tm)
```
If this was intended to be python only, I would suggest using a generator.

Even in C++ it might be useful to have a lazy iterator.

greglandrum replied:
I'm going to assume that this will move to C++ in the "near future", but doing it as a generator now certainly makes sense.

bp-kelley commented:
@greglandrum 

(1)  it is useful to ignore some nitrogen stereo swaps as they really don't change the conformation that much especially when using the  mmff94s variant where aniline nitrogens are planar due to time averaging.  Saves a power of 2 on the swaps as well.

(2) there is no 2

greglandrum commented:
@bp-kelley : point 1 makes sense. Can you provide some examples of N atoms that the RDKit considers (potential) stereocenters where the swaps can be ignored?

coleb commented:
Hi guys, late to the party as I've actually been on vacation during my vacation. Some general notes below: 

- The trick of trying the embedding is quite cute and cool, though I typically only do stereo enumeration before doing conformer generation so I wouldn't use it.

- Should we consider an options object as the number of options to the function appears to be growing large, and will likely grow larger. 

- Should this automatically de-duplicate isomers? For example, the following are the same and a trip through MolToSmiles would deduplicate them. 
F[C@@H](Cl)[C@@H](Cl)F
F[C@H](Cl)[C@H](Cl)F

- I find it more useful to generate a random subset of isomers instead of reordering the stereo centers. We could collapse maxNumCenters and randomizeOrderOfCenters into a single maxIsomers argument that can be documented as 'maxIsomers: the maximum number of isomers to yield, if the number of possible isomers is greater than maxIsomers, a random subset will be yielded. If 0, all isomers are yielded. Since every additional stereo center doubles the number of results (and execution time) it's important to keep an eye on this.'

- I definitely agree that bond stereo enumeration should be included, that's why I wanted the CIS/TRANS stuff in RDKit in the last release. :-)

- This code currently doesn't handle pseudo-stereo centers (stereo centers that only exist whenever other stereo centers are set that create asymmetry). For example: CCC=CC(=C(CC)CO)C=CCC has 4 unique isomers: 

![image](https://user-images.githubusercontent.com/127899/29625080-ba62c3ce-87f8-11e7-9606-9d57734a5a4d.png)

The way I work around that here is to recursively look for more stereo centers on an isomer once an isomer is created: https://gist.github.com/coleb/9d132e053d40cacb648e1c3c9bd6a625

The gist contains the current stereo enumeration code I've been using and the test cases I've found that deal with the pseudo-stereo center case. It's not full featured with all the options this implementation has. And coming up with a way to deal with pseudo-chiral and sampling a subset of stereo centers will be definitely be tricky. I could maybe bite off integrating this stuff together as my RDKit UGM hack day project, that would be fun.

greglandrum commented:
I'm also slow since my vacation was also really a vacation. I'm back now though.

@coleb : thanks for the suggestions. Most of it makes sense to me and I think integrating your work would make a lot of sense. Would it help if I got a start on implementing things like the options argument and left the more complex bit of integrating your changes for dependant stereochemistry to you?

coleb commented:
@greglandrum, yes, that would definitely help as there aren't many examples in the codebase I could base that off of yet.

greglandrum commented:
Alright. Will try and get something in tomorrow. Will ping you when the code is there.

greglandrum commented:
@coleb : the first pass at using an option class is now there.

bp-kelley commented:
@greglandrum I'll dig up some examples this week (hopefully)

greglandrum commented:
@bp-kelley: if you're talking about three-coordinate Ns, you can avoid the searching since the RDKit currently doesn't consider them as possible chiral centers:
```
In [2]: Chem.CanonSmiles('CC[N@](Cl)F')
Out[2]: 'CCN(F)Cl'
```

bp-kelley commented:
@greglandrum @coleb Are we sufficient for adding a first pass at this?  We can add additional options later.

As for nitrogens, the biggest offenders are aniline:

<img width="472" alt="aniline" src="https://user-images.githubusercontent.com/9888416/30274938-73328c58-96cd-11e7-980a-7831e6c66d67.png">

greglandrum commented:
The "first pass" argument is a reasonable one.
I think the current state of this PR is better than nothing (which is what's currently in the code).
As long as @coleb agrees I think it'd be fine to merge this (with a comment about known weaknesses, which I will add) and then add his enhancements when they are ready

coleb commented:
I would make at least one change as a first pass, just remove the `randomizeOrderOfCenters` option as it conflicts with how I would like to implement a subsetting mechanism. Also, I don't think APIs should call 'random.seed' as that has side effects outside the calling function. 

Another small change to consider is renaming from `GenerateAllStereoisomers` to `GenerateStereoisomers` as it won't necessarily generate 'all' of them, and shorter is usually better in naming. 

Hoping to make this my hack day next week!

greglandrum commented:
Those both make sense. I just removed the random option (I like having something like that in there, but we can add it back later if we want) and renamed to `GenerateStereoisomers()`.

bp-kelley commented:
@greglandrum @coleb Going to merge tonight if I don't hear from either of you.
